ugvc/mrd/srsnv_plotting_utils.py

```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        output_filename_beeswarm: str = None,
+        n_sample: int = 10_000,
+        plot_feature_importance: bool = True,
+        plot_beeswarm: bool = True,
+        feature_importance_kws: dict = None,
+        beeswarm_kws: dict = None,
+    ):
+        """Calculate and plot SHAP values for the model."""
+        feature_importance_kws = feature_importance_kws or {}
+        beeswarm_kws = beeswarm_kws or {}
+        # Define model, data
+        k = 0  # fold_id
+        model = self.models[k]
+        X_val = self.data_df[self.data_df["fold_id"] == k]
+        # Get SHAP values
+        logger.info("Calculating SHAP values")
+        shap_values, X_val, y_val = self._shap_on_sample(  # pylint: disable=unused-variable
+            model, X_val, n_sample=n_sample
+        )
+        logger.info("Done calculating SHAP values")
+        # SHAP feature importance
+        mean_abs_SHAP_scores = pd.Series(
+            np.abs(shap_values[:, 1, :-1] - shap_values[:, 0, :-1]).mean(axis=0), index=X_val.columns
+        ).sort_values(ascending=False)
+        mean_abs_SHAP_scores.to_hdf(self.output_h5_filename, key="mean_abs_SHAP_scores", mode="a")
+        # Plot shap scores
+        if plot_feature_importance:
+            self.plot_SHAP_feature_importance(
+                shap_values, X_val, output_filename=output_filename_importance, **feature_importance_kws
+            )
+        if plot_beeswarm:
+            self.plot_SHAP_beeswarm(shap_values, X_val, output_filename=output_filename_beeswarm, **beeswarm_kws)
+
+    def _get_trinuc_stats(self, q1: float = 0.1, q2: float = 0.9):
+        data_df = self.data_df.copy()
+        data_df["is_cycle_skip"] = data_df["is_cycle_skip"].astype(int)
+        trinuc_stats = data_df.groupby(["trinuc_context_with_alt", "label", "is_forward", "is_mixed"]).agg(
+            median_qual=("qual", "median"),
+            quantile1_qual=("qual", lambda x: x.quantile(q1)),
+            quantile3_qual=("qual", lambda x: x.quantile(q2)),
+            is_cycle_skip=("is_cycle_skip", "mean"),
+            count=("qual", "size"),
+        )
+        trinuc_stats["fraction"] = trinuc_stats["count"] / self.data_df.shape[0]
+        trinuc_stats = trinuc_stats.reset_index()
+        trinuc_stats["is_forward"] = trinuc_stats["is_forward"].astype(bool)
+        return trinuc_stats
+
+    def _get_trinuc_with_alt_in_order(self, order: str = "symmetric"):
+        """Get trinuc_with_context in right order, so that each SNV in position i
+        has the opposite SNV in position i+96.
+        E.g., in position 0 we have A[A>C]A and in posiiton 96 we have A[C>A]A
+        """
+        assert order in {"symmetric", "reverse"}, f'order must be either "symmetric" or "reverse". Got {order}'
+        if order == "symmetric":
+            trinuc_ref_alt = [
+                c1 + r + c2 + a
+                for r, a in (("A", "C"), ("A", "G"), ("A", "T"), ("C", "G"), ("C", "T"), ("G", "T"))
```

Comment:
Shouldn't the last one be "C", "A"? In fact, I think this is the line you wanted:
`                for r, a in (("A", "C"), ("A", "G"), ("A", "T"), ("C", "A"), ("C", "G"), ("C", "T"))`

Reply:
I might not have understood something but it looks like an anomaly